### ISSUE TITLE
Coloring scheme

### ISSUE THREAD

**Issue by idomic** [ploomber/sklearn-evaluation#243]
We can come up with a unique/prettier coloring scheme for our plots.
This can be unified across the different plots and we can also add a @Made by ploomber tag.
For instance, gigi plot on matplotlib (looks nicer than the default one)

**Comment by yafimvo** [comment]
I reviewed matplotlib [coloring and styling API](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot.html#examples-using-matplotlib-pyplot-plot), also ggplot, plotly, seaborn, yellowbrick, and some other plotting libraries.

While some of these are more visually appealing than others, most of them have quite complicated UX. 
YB offers an easy color setup but their `rcmod` (allows to modify matplotlib rcParams) doesn't work very well.

I think we have an opportunity to simplify things by making informative, branded, good-looking plots out of the box.

---

First, by default, we can create better-looking plots. something like these:

![image](https://user-images.githubusercontent.com/103026689/216285580-0cd08522-2f9d-478e-8cc1-1bbe1d56ee3c.png)

![image](https://user-images.githubusercontent.com/103026689/216286202-aef744cb-1368-4936-8440-0e32776e365f.png)


Second, to make it simple for our users and consistent with our current plots, I thought of the following API:

Modify the default theme from outside by changing specific parts of the plot (line colors, background, etc...)
** Maybe we can add modes (dark, classic, material, etc...)

```
plot.ROC.set_theme(line_colors=[...], background, )
```

The matplotlib styling API offers a varied range of properties, I'm not sure we need them in our cases, but we can use them as follows when applying a specific style to each plot (instead of using the `labels` list we have now)

```python
plot_1_style= dict({
    'color' : 'magenta',
    'label':'Enfield', 
    'linewidth':5, 
    'marker':'o', 
    'markerfacecolor':'blue', 
    'markersize':12
})

plot.ROC(fpr, tpr, styles = [plot_1_style, plot_2_style, ...] )
```

this is good when our users want to hide or mark a specific plot, e.g:
![image](https://user-images.githubusercontent.com/103026689/216299572-f6218d2e-ed80-4012-971b-5dc9447b9ee5.png)


@idomic @edublancas thoughts?

**Comment by edublancas** [comment]
Great overview!

We can do this in a few steps. The first one is have a consistent coloring scheme for all our plots. I remember reading in matplotlib (or seaborn) some explanation about how to choose colors (that are nice to colorblind people, among other things) and gave some examples of palettes with these properties. We can start with one of those as a base and modify some colors, but ensure we still respect the best practices.

The second part is how users will activate this. Let's look at seaborn and see how they do it, AFAIK, all their plots are customized to their theme, but the theme only comes out when calling seaborn (seaborn uses matplotlib under the hood); hence, it doesn't modify the global matplotlib configuration. I think we want the same thing, otherwise, we'll end up modifying the global state and any other non-sklearn-evaluation plotting calls will have the sklearn-evaluation theme.


Once this is done, we can discuss an API for customization.

**Comment by idomic** [comment]
That's great @yafimvo!
I think we should def. take it in two steps as mentioned above.

### Step 1
1. Design and decide on the color palette. 
2. Add it to the implementation as part of the schemes (I liked the dark mode idea as well).

### Step 2
1. Add options to customize it via overriding some of the matplotlib.
2. Add advanced features like marking the specific line we're looking at.

**Comment by edublancas** [comment]
> Material UI palette

this looks good!

> Make the plot frame L shape and not a square

I also like this. We should also style the grid (display it with: `ax.grid()`). In most plots, this helps visualization. here's an example from [seaborn](https://seaborn.pydata.org/examples/errorband_lineplots.html#timeseries-plot-with-error-bands)

> Remove the legend outside of the plot

@gtduncan was working on this for ROC, I believe. He found that matplotlib doesn't offer a simple way to place the legend outside the plot, so sounds like we'd have to do it manually

---

I think this would still modify matplotlib's global state:

```python
import matplotlib as mpl

_material_ui_palette = ['#00B0FF', '#F50057', '#00C853', '#D500F9', '#F44336', '#FF9100', '#00BFA5', '#651FFF']
mpl.rcParams['axes.prop_cycle'] = mpl.cycler(color=_material_ui_palette )
```

But I think we can get way by storing the original values and restoring them after plotting (we can use a context manager):

```python
from sklearn_evaluation.style import tmp_theme, apply_theme

with tmp_theme():
    plt.plot([1,2,3])
```

And to simplify, we could wrap it in a decorator:

```python
@apply_theme
def plot():
    ...
```

Then we could apply it to our different plots.


> The "hard" part would be updating all our existing plot tests and if we move out the legend - we need to check how it affected our docs (since the image is wider than before)


Yes. I think let's implement the decorator pattern, then, we can create another issue to apply it to different plots and check if everything looks good. I'd say let's leave the legend placement out of this iteration and do it for v2.

**Comment by edublancas** [comment]
a good source of inspiration is R's ggplot package: https://ggplot2.tidyverse.org/

**Comment by edublancas** [comment]
FYI: https://matplotlib.org/stable/tutorials/introductory/customizing.html#temporary-rc-settings

**Comment by yafimvo** [comment]
> @gtduncan was working on this for ROC, I believe. He found that matplotlib doesn't offer a simple way to place the legend outside the plot, so sounds like we'd have to do it manually

this one should do the job. What do you mean manually?
```python
plt.legend(loc=(1.05, 0.5))
```

@gtduncan @edublancas

**Comment by gtduncan** [comment]
That did work for me partially when I tried it, but the figure for each plot wasn't expanding so the resulting plots that were produced had the legend cut off pretty much entirely. I looked into some solutions and nothing seemed to fix it for me

**Comment by edublancas** [comment]
> this one should do the job (I'm not sure how it will work on custom plot size). What do you mean manually?

By manually I mean something like what you described (`plt.legend(loc=(1.05, 0.5))`) - manually placing the box by setting the coordinates.

As @gtduncan mention, I think the main issue is that if we place the box by setting the coordinates is that it'll cramp the plot making it harder to read. Perhaps we can put it below?

Ideally we could use the [`best`](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.legend.html#matplotlib-pyplot-legend) setting in matplotlib and only move it outside the plot when there is a lot of overlap ([here's the implementation](https://github.com/matplotlib/matplotlib/blob/6a9a07155c0e7f91c20dd4c7e280198ec652c4ae/lib/matplotlib/legend.py#L1121))

since this will require a lot of extra work, I suggest we deal with this in the second iteration
